### ISSUE TITLE
Translating this query requires the SQL APPLY operation, which is not supported on SQLite.

### ISSUE THREAD

**Issue by xueshaoyu** [dotnet/efcore#26684]
## File a bug

Remember:

* Please check that the [documentation](https://docs.microsoft.com/ef/) does not explain the behavior you are seeing.
* Please search in both [open](https://github.com/dotnet/efcore/issues) and [closed](https://github.com/dotnet/efcore/issues?q=is%3Aissue+is%3Aclosed) issues to check that your bug has not already been filed.

### Include your code

To fix any bug we must first reproduce it. To make this possible, please attach a small, runnable project or post a small, runnable code listing that reproduces what you are seeing.

It is often impossible for us to reproduce a bug when working with only code snippets since we have to guess at the missing code. 

Use triple-tick code fences for any posted code. For example:

```C#
  return await _userRepository.Include(u => u.Roles, false)
                                            .ThenInclude(u => u.Securities)
                                            .Where(u => u.Id == userId)
                                            .SelectMany(u => u.Roles
                                            .SelectMany(u => u.Securities).Distinct())
                                            .ToListAsync();
```

### Include stack traces

```
System.InvalidOperationException: Translating this query requires the SQL APPLY operation, which is not supported on SQLite.
   at Microsoft.EntityFrameworkCore.Sqlite.Query.Internal.SqliteQueryTranslationPostprocessor.ApplyValidatingVisitor.VisitExtension(Expression extensionExpression)
   at System.Linq.Expressions.Expression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at Microsoft.EntityFrameworkCore.Sqlite.Query.Internal.SqliteQueryTranslationPostprocessor.ApplyValidatingVisitor.VisitExtension(Expression extensionExpression)
   at System.Linq.Expressions.Expression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)

```

### Include provider and version information

EF Core version: 6.0
Database provider:Microsoft.EntityFrameworkCore.Sqlite
Target framework: NET 6.0
Operating system:
IDE:vs2022

**Comment by ajcvickers** [comment]
@xueshaoyu As the message says, the SQL needed to translate this query is not supported on SQLite. If the SQLite database system adds support, then we can add corresponding support in EF Core. Until then, this kind of query cannot be translated when using the SQLite provider.

**Comment by vanpyrzericj** [comment]
@ajcvickers it would be great to get some guidance on what exactly in this query is unsupported (or how to determine that, on our own). I'm running into a similar issue, with a pretty straightforward query (in my opinion).

Neither my query, nor the one @xueshaoyu posted here _seem_ to overlap with any of the stated SQLite limitations that are documented [here](https://learn.microsoft.com/en-us/ef/core/providers/sqlite/limitations). I could (hope) I'm wrong, but without knowing exactly what the obstacle is here, its difficult to triage what needs to be adjusted in the query in order to ensure compatibility. Additionally, in my case at least, these queries were running fine in my unit tests (which use SQLite) until I upgraded from EF Core 6 to v8. That is when these particular tests stopped working.

**Comment by Ben555555** [comment]
@vanpyrzericj 
I'm facing similar issues and also can't find a clear pattern:

Works:
```
dbContext.Users.Select(user => {
RoleIds = dbContext.UserRoles.Where(userRole => userRole.UserId == user.Id).Select(userRole => userRole.RoleId).ToList()
}).ToList()
```

Does not work:
```
dbContext.Users.Select(user => {
RoleIds = dbContext.UserRoles.Select(userRole => userRole.RoleId).ToList()
}).ToList()
```

Works:
```
dbContext.Users.Select(user => {
Rights = user.UserRoles
    .SelectMany(userRole => userRole.Role.RoleRights
    .Select(roleRight => roleRight.Right.Key))
    .ToList()
}).ToList()
```

Does not work:
```
dbContext.Users.Select(user => {
Rights = user.UserRoles
    .SelectMany(userRole => userRole.Role.RoleRights
    .Select(roleRight => roleRight.Right.Key))
    .Distinct()
    .ToList()
}).ToList()
```

Does not work:
```
dbContext.Users.Select(user => {
Rights = dbContext.RoleRights.Where(roleRight => roleRight.Role.UserRoles
    .Any(userRole => userRole.UserId == user.Id))
    .Select(roleRight => roleRight.RightKey)
    .ToList()
}).ToList()
```


So whatever I try, nothing works. I wanted to switch from InMemory to SQLite but SQLite has even more limitations...

**Comment by LJFloor** [comment]
Here is what I found in my scenario. It is a different scenario than OP, but somebody googling for this might find this useful.

This LINQ query works. It selects albums along with their artists and tracks, including each track’s artist list:
```csharp
var result = await dbContext.Albums
    .OrderBy(a => a.Name)
    .Select(album => new
    {
        album.Id,
        album.Name,
        Artists = album.Artists.Select(artist => artist.Name),
        Tracks = album.Tracks.Select(track => new
        {
            track.Id,
            track.Title,
            track.TrackNumber,
            track.TotalSeconds,
            track.Year,
            Artists = track.Artists.Select(artist => artist.Name),
        })
    })
    .ToArrayAsync();
```

When I try to also add an album to the track (I used this for navigation in the frontend), it errors out:
```csharp
var result = await dbContext.Albums
    .OrderBy(a => a.Name)
    .Select(album => new
    {
        album.Id,
        album.Name,
        Artists = album.Artists.Select(artist => artist.Name),
        Tracks = album.Tracks.Select(track => new
        {
            track.Id,
            track.Title,
            track.TrackNumber,
            track.TotalSeconds,
            track.Year,
            Artists = track.Artists.Select(artist => artist.Name),
            Album = new
            {
                album.Id,   // ❌ Problem: accessing outer variable in nested projection
                album.Name
            }
        })
    })
    .ToArrayAsync();
```

## Why this failed
Turns out, EF Core with the SQLite provider cannot translate a projection where a nested subquery tries to reuse the outer variable **album** inside a deeper projection. For me, this threw:
```
Translating this query requires the SQL APPLY operation, which is not supported on SQLite.
```

TL;DR: Once you enumerate into `Track`, you cant access `Album`.

## Fix

My fix was to use the exsiting property `track.Album`. So instead of

```csharp
Album = new
{
    album.Id,
    album.Name
}
```

I now use

```csharp
Album = new
{
    track.Album.Id,
    track.Album.Name
}
```
